### ISSUE TITLE
Candle Inference ~8.5x Slower Than PyTorch on CPU

### ISSUE THREAD

**Issue by msminhas93** [huggingface/candle#2877]
## Issue

Inference with the `sentence-transformers/all-MiniLM-L6-v2` model using `candle-transformers` (v0.8.4) is significantly slower (approximately 8.5 times) compared to the standard PyTorch implementation using transformers and PyTorch. The average time per batch (size 4) is around 122.30 ms in Candle versus ~14.34 ms in PyTorch.

## Environment

*   **Candle Version**: 0.8.4 (using `candle-core`, `candle-nn`, `candle-transformers`)
*   **Candle Features**: `cuda`, `cudnn`, `mkl`
*   **CUDA Version**: 12.8 (V12.8.93)
*   **Model**: `sentence-transformers/all-MiniLM-L6-v2` 
*   **Batch Size**: 4
*   **Hardware**: i9-13900HX
*   **Operating System**: Windows 11

## Profiling Data (Batch Size = 4)

Time taken for the forward pass.

| Language | Operation     | Mean (ms) | Std (ms) | Min (ms) | Max (ms) |
| :------- | :------------ | :-------- | :------- | :------- | :------- |
| Python   PyTorch | Inference (ms) | 14.34     | 3.15     | 8.00     | 120.78   |
| Rust  Candle   | Inference (ms) | 122.30    | 43.38    | 18.00    | 319.00   |

## Expected Result

Inference time should be closer to the PyTorch performance, ideally around 14-20 ms on CPU. I am observing Candle GPU performance better than PyTorch vanilla. 

## Actual Result

The average inference time with Candle is ~122.30 ms per batch, with significant variance (standard deviation: 43.38 ms). This is much slower than expected and slower than previous reports comparing Candle and Python implementations for this model.

**Comment by EricLBuehler** [comment]
Hi @msminhas93! Can you please make a flamegraph using https://github.com/flamegraph-rs/flamegraph?

**Comment by msminhas93** [comment]
Hey Eric! Here is the flamegraph. Seems like bulk of it is used by rayon. Since I'm not using any rayon in my code it seems like candle is using it. I'm not that great with flamegraphs yet so my observation could be incorrect. 
![Image](https://github.com/user-attachments/assets/4d7a90ea-f729-4975-ab3d-2fb24579cbca)

**Comment by EricLBuehler** [comment]
Hi @msminhas93, thanks! Can you please send the original .svg file for easier inspection?

**Comment by BerserkerMother** [comment]
Hi @msminhas93 
how are you building you binary?

**Comment by msminhas93** [comment]
@BerserkerMother the normal cargo build --release 
with debug set to 1 for flamegraph
[profile.release]
debug = 1

**Comment by LaurentMazare** [comment]
You may want to check whether your simd instructions were properly detected, e.g. by printing `candle::utils::with_avx()` for x86.

**Comment by BerserkerMother** [comment]
> You may want to check whether your simd instructions were properly detected, e.g. by printing `candle::utils::with_avx()` for x86.

Shouldn't he build with RUSTFLAGS="-C target-feature=+avx" for this?

**Comment by LaurentMazare** [comment]
The best is probably to have the same `.cargo/config.toml` at the root of the project as in the main candle repo [here](https://github.com/huggingface/candle/blob/main/.cargo/config.toml).

**Comment by msminhas93** [comment]
I tried adding the config.toml but doesn't seem to help. AVX is enabled. 

println!("AVX enabled: {}", utils::with_avx());
AVX enabled: true

Rayon does seem to be the bulk of the flamegraph. I don't get it.
The forward pass doesn't use any rayon.

```rust
    pub fn forward(
        &self,
        input_ids: &Tensor,
        token_type_ids: &Tensor,
        attention_mask: Option<&Tensor>,
    ) -> Result<Tensor> {
        let _enter = self.span.enter();
        let embedding_output = self.embeddings.forward(input_ids, token_type_ids)?;
        let attention_mask = match attention_mask {
            Some(attention_mask) => attention_mask.clone(),
            None => input_ids.ones_like()?,
        };
        let dtype = embedding_output.dtype();
        // https://github.com/huggingface/transformers/blob/6eedfa6dd15dc1e22a55ae036f681914e5a0d9a1/src/transformers/models/bert/modeling_bert.py#L995
        let attention_mask = get_extended_attention_mask(&attention_mask, dtype)?;
        let sequence_output = self.encoder.forward(&embedding_output, &attention_mask)?;
        Ok(sequence_output)
    }
```

**Comment by msminhas93** [comment]
I enabled mkl and that helped some to reduce the mean from 122.3ms to 83.3ms. This is the flamegraph of mkl enabled run.

![Image](https://github.com/user-attachments/assets/a2159fac-2d80-4a8c-b925-de98dcac6e16)

**Comment by LaurentMazare** [comment]
Flamegraphs are pretty hard to look at as indeed lots of things will be hidden by rayon. Instead could you use the tracing functionality, e.g. something similar to [this](https://github.com/huggingface/candle/blob/15ed0b11cef868bbcf58dfa87c796db84fdcaff2/candle-examples/examples/bert/main.rs#L24), this should provide a trace that is compatible with perfetto.
Re mkl, could you report the min and max time taken in this case. In your original stats, it seemed that one of the forward could have been much slower than the rest with may be indicative of some initialization work taking place.

**Comment by BerserkerMother** [comment]
> Flamegraphs are pretty hard to look at as indeed lots of things will be hidden by rayon. Instead could you use the tracing functionality, e.g. something similar to [this](https://github.com/huggingface/candle/blob/15ed0b11cef868bbcf58dfa87c796db84fdcaff2/candle-examples/examples/bert/main.rs#L24), this should provide a trace that is compatible with perfetto. Re mkl, could you report the min and max time taken in this case. In your original stats, it seemed that one of the forward could have been much slower than the rest with may be indicative of some initialization work taking place.

I didn’t know that. Thank you very much! Do you think we can improve the book to include such things? I was stuck trying to find execution times with Flamegraph, but now I know there’s a much better way.

**Comment by LaurentMazare** [comment]
Yeah the book is not in a very good state and lags quite a bit compared to the current candle stuff + probably doesn't focus on the most relevant things. Maybe some rewrite effort is needed there but it's hard to ensure that it stays current. @greenrazer maybe that's something that you would be interested in looking at ?

Also just to mention another trick, you may want to set `RAYON_NUM_THREADS=1` for performance investigation. Obviously it will impact performance by disabling multi-threading but for some of the kernels (at least gemm I think) it will also turn off rayon entirely so could result in much nicer flamegraphs.

**Comment by BerserkerMother** [comment]
@LaurentMazare I'll be happy to help and learn in the way.

**Comment by msminhas93** [comment]
Rayon is adding overhead for inference at least for this model and mkl version. These are the numbers for different threads mkl vs no mkl and python. I'll look at the chrome profiling and post my findings soon. The data has 1904 baches at batch size of 4. 

Comparison of 'Inference (ms)' Statistics:

|                                             |   count |    mean |     std |      min |   25% |     50% |     75% |     max |
|:--------------------------------------------|--------:|--------:|--------:|---------:|------:|--------:|--------:|--------:|
| profiling_results_rust_mkl_1_threads.csv    |    1904 | 53.9296 | 14.6833 | 15       |    46 | 53      | 59      | 159     |
| profiling_results_rust_mkl_4_threads.csv    |    1904 | 60.9974 | 18.3994 | 15       |    51 | 59      | 68      | 196     |
| profiling_results_rust_mkl_8_threads.csv    |    1904 | 73.8015 | 22.3596 | 17       |    57 | 73      | 87      | 223     |
| profiling_results_rust_no_mkl_1_threads.csv |    1904 | 64.583  | 15.0718 | 14       |    54 | 65      | 74      | 135     |
| profiling_results_rust_no_mkl_4_threads.csv |    1904 | 50.7468 | 11.8582 | 11       |    42 | 51      | 58      | 108     |
| profiling_results_rust_no_mkl_8_threads.csv |    1904 | 50.4695 | 11.7823 | 11       |    42 | 51      | 58      | 110     |
| profiling_results_python.csv                |    1904 | 14.3405 |  3.1515 |  8.00014 |    13 | 14.0002 | 15.0008 | 120.781 |

**Comment by msminhas93** [comment]
For CPU ort https://github.com/pykeio/ort/blob/main/examples/sentence-transformers/semantic-similarity.rs speed for the same model faster than pytorch version. 
| Metric         | Mean      | Std       | Min       | Max       |
|----------------|-----------|-----------|-----------|-----------|
| Inference (ms) | 8.498950  | 2.709629  | 2.000000  | 67.000000 |

**Comment by aayushshah27894** [comment]
I’m evaluating a switch from my existing Python inference stack to Rust for an edge-deployment scenario. The target node is an IBM x3650 M4 (Intel Xeon, 12 cores / 24 threads, 16 GB RAM) on Windows Server 2016, but the service will only get a fraction of those resources because other workloads share the machine.

Based on your experience working on this project @msminhas93, do the latency-and-footprint gains you’ve seen with Rust warrant the migration effort in this context?

**Comment by D0NMEGA** [comment]
Hi @LaurentMazare - planning a CPU SIMD matmul fast-path PR for the f32 path inside impl Map2 for MatMul (candle-core/src/cpu_backend/mod.rs), targeting the FFN-up shape that MiniLM-L6 hits ([512, 384] x [384, 1536]). Will include before/after criterion benchmark numbers and keep the gemm v0.19.0 pin intact. OK to claim?

**Comment by D0NMEGA** [comment]
Quick update on the SIMD claim above: tried a naive AVX2 8x8 tile fast-path on the MiniLM-L6 shapes ([512,384]x[384,1536] FFN-up and the QKV / FFN-down companions) and measured 7-17x SLOWER than the existing gemm v0.19.0 dispatch across every benchmarked shape on Intel Mac. gemm wins decisively here; would need to port its multi-level cache-blocking + register-tiling strategy to compete. Releasing the claim - happy to leave #2877 open for someone with that kernel-tuning bandwidth. Bench numbers below for the record.

```
Shape                              | gemm baseline   | naive-SIMD median | speedup
-----------------------------------|-----------------|-------------------|---------
minilm_qkv [512,384]x[384,384]     | 504 us          | 2.95 ms           | 0.17x
minilm_ffn_up [512,384]x[384,1536] | 1.83 ms         | 13.67 ms          | 0.13x
minilm_ffn_down [512,1536]x[1536,384] | 2.01 ms      | 13.18 ms          | 0.15x
square_512                         | 752 us          | 6.13 ms           | 0.12x
square_1024                        | 5.95 ms         | 76.87 ms          | 0.077x
linear_large (BERT FFN)            | 353 ms          | 5.95 s            | 0.059x
```

**Comment by ivarflakstad** [comment]
Hey!
Now that metal is at a decent place (~90% of mlx, equal to llama.cpp on quantized models) I've started investigating the cpu performance. I started with quantized models, since llama.cpp had ridicolous performance compared to us.
Like mentioned here it really is rayon, or more precisely the way we use rayon, that is the first big bottleneck. Too many competing threads. This hides the true performance gain made by improving our kernels (like cpu flash attn work by @DrJesseGlass, @michaelfeil, and @EricLBuehler, or improved striding and convolutions by @slckl, etc) - which incidentally is the exact same situation as the recent improvements in metal.

Anyways - both Qwen3 0.6B Q4_K_M and Q8_0 were at ~30 tok/s on CPU on my machine. They are now at ~170-180 tok/s.
I'm still exploring so haven't made any PRs yet. I'll get to it soon.

**Comment by ivarflakstad** [comment]
@D0NMEGA unlucky choice of optimization - gemm is kinda the only thing in the candle cpu backend that is truly optimized.
IIRC there is room for improvement on small matrices (@slckl I remember you mentioned something about this? Compared it with faer perhaps?), but other than that it's pretty close to SOTA.

**Comment by ivarflakstad** [comment]
Llama-3.2-1B producing 200 tokens on CPU in full precision (with Accelerate enabled on all):

| Framework | TTFT (ms) | tok/s |
| ---- | ----- | ----- |
| candle| 102.9 | 35 |
| candle@0.8.4 | 110.9 | 31 |
| pytorch | 109.6 | 34 |
| mlx | 175.2 | 39 |

mlx just uses more of Accelerate/LAPACK than the rest. Not much more to it.

Edit: This is on the M4 Max. With Accelerate disabled we're on par with mlx. I have not profiled or optimized for other CPU architectures yet as this is the machine I have at hand.

**Comment by slckl** [comment]
> [@D0NMEGA](https://github.com/D0NMEGA) unlucky choice of optimization - gemm is kinda the only thing in the candle cpu backend that is truly optimized. IIRC there is room for improvement on small matrices ([@slckl](https://github.com/slckl) I remember you mentioned something about this? Compared it with faer perhaps?), but other than that it's pretty close to SOTA.

faer has both `gemm`, which we already use, and [`nano-gemm`](https://lib.rs/crates/nano-gemm) for small matrices. I did experiment with it a little bit, but I don't recall seeing any surprising gains (it seemed to be roughly the same for the stuff I looked at), but my benchmarks at the time were extremely limited.
